packages/frontend/src/state/lp/hooks.ts

```diff
@@ -0,0 +1,243 @@
+      const vaultBefore = await getVault(vaultId)
+      if (!contract || !address || !squeethPoolContract || !vaultBefore || !vaultBefore.shortAmount || !vaultBefore.collateralAmount) return null
+      
+      const mintWSqueethAmount = fromTokenAmount(squeethToMint, OSQUEETH_DECIMALS)
+
+      // Calculate prices from ticks
+      const { tick, tickSpacing } = await getPoolState(squeethPoolContract)
+      const lowerTick = nearestUsableTick(lowerTickInput, Number(tickSpacing))
+      const upperTick = nearestUsableTick(upperTickInput, Number(tickSpacing))
+      const sqrtLowerPrice = new BigNumber(TickMath.getSqrtRatioAtTick(lowerTick).toString()).div(x96)
+      const sqrtUpperPrice = new BigNumber(TickMath.getSqrtRatioAtTick(upperTick).toString()).div(x96)
+      const squeethPrice = isWethToken0 ? new BigNumber(1).div(new BigNumber(TickMath.getSqrtRatioAtTick(Number(tick)).toString()).div(x96).pow(2))
+                                            : new BigNumber(TickMath.getSqrtRatioAtTick(Number(tick)).toString()).div(x96).pow(2)
+      const sqrtSqueethPrice = squeethPrice.sqrt()
+
+      let collateralToLp
+      if (sqrtUpperPrice.lt(sqrtSqueethPrice)) {
+        // All weth position
+        console.log("LPing an all WETH position is not enabled, but you can rebalance to this position.")
+        return
+      } else if (sqrtSqueethPrice.lt(sqrtLowerPrice)) {
+        // All squeeth position
+        collateralToLp = new BigNumber(0)
+      } else {
+        // Lx = x * (sqrtSqueethPrice * sqrtUpperPrice) / (sqrtUpperPrice - sqrtSqueethPrice)
+        // y = Lx * (sqrtSqueethPrice - sqrtLowerPrice)
+        const liquidity = mintWSqueethAmount.times(sqrtSqueethPrice.times(sqrtUpperPrice)).div(sqrtUpperPrice.minus(sqrtSqueethPrice))
+        collateralToLp = sqrtUpperPrice.lt(sqrtSqueethPrice) ? liquidity.times(sqrtUpperPrice.minus(sqrtLowerPrice))
+                              : sqrtSqueethPrice.lt(sqrtLowerPrice) ? new BigNumber(0)
+                              : liquidity.times(sqrtSqueethPrice.minus(sqrtLowerPrice))
+      }
+      
+      const amount0New = isWethToken0 ? collateralToLp : mintWSqueethAmount
+      const amount1New = isWethToken0 ? mintWSqueethAmount : collateralToLp
+      const amount0Min = amount0New.times(new BigNumber(1).minus(slippage)).toFixed(0)
+      const amount1Min = amount1New.times(new BigNumber(1).minus(slippage)).toFixed(0)
+
+      const collateralToWithdraw = fromTokenAmount(withdrawAmount, OSQUEETH_DECIMALS)
+      const ethIndexPrice = toTokenAmount(index, 18).sqrt()
+      const vaultShortAmt = fromTokenAmount(vaultBefore.shortAmount, OSQUEETH_DECIMALS)
+      const vaultCollateralAmt = fromTokenAmount(vaultBefore.collateralAmount, WETH_DECIMALS)
+      
+      // Calculate collateralToMint
+      const collateralToMint = (new BigNumber(collatRatio).times((vaultShortAmt.plus(mintWSqueethAmount)).times(normFactor).times(ethIndexPrice).div(INDEX_SCALE)))
```

Comment:
We need to take into account the LP composition as our system sees it.

Our system treats oSQTH in LPs as an ETH amount equal to oSqthInLp * (EthPrice / 10000) * NF

So we need to add back the ETH in LP as well as the oSQTH value in ETH terms in the LP in this manner, not the eth value based on squeeth price.

Personally also got thrown off by the .minus with all the collateral together (some added) vs just a set of .minus(vaultCollateralAmt).plus(collateralToWithdraw).minus(collateralToLP)...etc, but it is right, just read it wrong.

Reply:
Fixed